Add pr/review guidelines (#456)
kokkos/kokkos-core-wiki: docs/source/developer-guides/prs-and-reviews.rst

> Consolidation:
> ==============

Reviewer: ```suggestion
PRs and Reviews
===============
```

> Overarching Goals of good Review
> --------------------------------
> 
> The goal of a review is to ensure that the proposed change is useful and maintainable over the long run.

Reviewer: ```suggestion
The goal of a review is to ensure that the proposed change is useful and maintainable over the long run. Submitters should consider and reviewers should evaluate the below criteria.
```

We can just cut this subheading directly under the heading.

PR Description
---------------

> - Have a meaningful title: it's easier when creating the changelog or when searching through old PR

Reviewer 2: ```suggestion
- Have a meaningful title: it's easier when creating the changelog or when searching through old PRs
```

- Motivate why we should merge the PR: adding/changing code risks introducing a new bug. IMO one person asking for a nice to have feature doesn't qualify.

> - Explain the PR does in the description: it makes it easier/faster to review

Reviewer 2: ```suggestion
- Explain what the PR does in the description: it makes it easier/faster to review
```

> - Make the PR as small as possible: it's much easier to review five 200 lines PR than one 1000 lines PR

Reviewer 2: ```suggestion
- Make the PR as small as possible: it's much easier to review five 200-line PRs than one 1000 lines PR
```

- If a PR is known to create conflict with other active PRs, try to coordinate, and link to each other

>   - if appropriate explain desired review/merge order 

Reviewer 2: ```suggestion
  - if appropriate, explain the desired review/merge order 
```

Reviewer 3: - In the case of conflicting PR, would this be too much to ask non-team members to manage / understand, particularly given the relatively great likelihood of some number of PR for View (for instance) having merge conflicts?
- Instead, would this make sense:  first, ask the potential PR author to check the PR queue to see if their desired change is already reflected in an existing PR; if not, then create a new PR with the (focused / "surgical" / single) desired change; if so, participate in the review of the existing PR, thereby helping move PR through the queues, and formulate ideas for future changes

> - For complex changes that will require multiple PR, create an issue to keep track of what has been done and what's left to do.

Reviewer 2: ```suggestion
- For complex changes that will require multiple PRs, create an issue to keep track of what has been done and what's left to do.
```

Reviewer 3: - For the issue "TODO" list, are you thinking something like [5674](https://github.com/kokkos/kokkos/issues/5674), with PR, and PR items to tick off upon completion?
- Would it be OK to provide this issue as an example of what you're talking about?

- Is the PR focused on a single bugfix or feature?

>   - consider moving self contained changes to separate PRs 

Reviewer 3: - I know what you mean here, but I'm not sure it would be clear to a wider audience
- Instead, could you say something like, "... please propose one, and only one focused / surgical change in one PR.  It is fine to submit multiple, focused PR to address a larger issue ..."

Public Interfaces
-----------------

> - are there comprehensive tests

Reviewer 3: - How does one know if the proposed tests are comprehensive?
- Can we firm up the concept, such that we give PR author a clear, actionable set of ideas on the minimum set of tests needed for a focused PR to pass muster?

Author: Don't think so, not without writing overly much and then having to constantly say: yeah ok this aspect was not covered in our detailed list because its impossible to make a detailed list since what comprehensive means may depend on the functionality tested.

Reviewer 3: Maybe split the difference and say something along the lines of "at least test coverage for the most obvious modes of failure" ...?

Author: No that is not what I mean: we need to test all ways something could go wrong.

- does the interface meet the use case - and for that matter is there a use case description

>   - is the usage intuitive

Reviewer 3: - If new, is the Use Case detailed in an Issue?
- Where do we expect a new Use Case to be described?  Same question for extending and/or modifying existing Use Cases

>   - do we really want this as a public interface, and maintain it?

Reviewer 3: - Isn't that something to be evaluated before getting to the PR creation stage?  Which is to say, something one would argue in an issue?

Reviewer 3: hmmm .... OK.  Maybe this is something we can / should mention explicitly in the "Contributing to Kokkos" entry in the Documentation.

- is the interface API consistent with existing ones?
  - e.g. if everything else takes execution space instances as the first argument, don't make it the last argument

>   - does the naming style match other Kokkos APIs

Reviewer 3: - Are there guidelines / expectations for naming style?

Author: not written down right now.

Reviewer 3: Here's your opportunity ; P

Author: But not part of this thing.

- are the interface semantics consistent with existing ones?
  - if everything else (or at least the majority) of interfaces taking an execution space instance argument is async, then new interfaces taking one should be too

> - For any routine that's going to launch any parallel work, does it have an overload that takes an execution space instance, and does that overload use the instance for all parallel work?

Reviewer 3: - Here, are you really asking a question about the PR author having provided all reasonable and necessary overload sets?  If so, address that point directly.
- Or, if your point is broader, and you're simply saying that parallel execution requires an execution space instance, try to simplify the statement

Author: I am not asking for all reasonably overload sets: we never want all things one could reasonably provide. In this case I want to very explicitly pointing out that stuff which does parallel work has to have an overload taking an instance - but there can be overloads which don't.

> - if the functionality is similar to an ISO C++ functionality, is the interface and behavior similar, and in places where it's not is it a conscious decision

Reviewer 3: - Does the "conscious decision" clause mean that there has to be a substantive, defensible reason for departing from ISO C++ ?

Author: yes that is what I am implying: you need to have thought about it - and I assume I don't need to point out that if you thought about and came to the conclusion that this is a dumb and unnecessary deviation, that you shouldn't have done this ...

- is there a corresponding API documentation PR
- Are the corner cases handled (works correct, won't compile, detected at run time)?
- Do the C++ defaulted functions do the right thing (including if needed to be marked with KOKKOS_DEFAULTED_FUNCTION)?

Internal Implementation
-----------------------

> - is the implementation style consistent with the rest of Kokkos

Reviewer 3: - Does the consistency thing refer to the theme you've previously mentioned about "self similarity?"  If so, this PR creation / reviewer guide might be a good place to roll the concept out more fully, using the term throughout  ...?

Author: no it has nothing to do with self similarity

- is there unnecessary code duplication
  - in particular: is code that can be shared across backends shared?

> - did it go in the right sub-part of Kokkos (core, algorithms, containers etc.)

Reviewer 3: - Wordsmithing alert:  instead of "sub-part," component?

- are there debug checks we should add?
  - like do the arguments make sense together etc.

> - are implementation details accidentally exposed?

Reviewer 3: - Is the implementation vs. API line generally bright and clear to potential PR authors?

Author: Not necessarily but I also don't think we should need to write 4 pages to explain it here (or generally) these are common C++ software engineering practices.

> - is there unnecessary fencing

Reviewer 3: - Would excessive fencing be evident to the average PR author?  If not, maybe add a rule of thumb on this point.

Author: I hope that someone reviews who does understand where fencing goes - i.e. when is synchronization necessary and when not. Generally I expect of every team member that they understand this, and I would consider it a necessary prerequisit skill for hiring someone for the team to do coding.

> - are there unnecessary allocations/deallocations

Reviewer 3: - Same question as above, would excessive allocation / deallocation be obvious to the average PR author?  If not, short rule of thumb to help author know when they're getting into trouble?

- Avoiding tangle of inclusions: are headers/files including only what is needed?
- Is the code expressing intent clearly (choosing expressive names for variables, functions, etc)?
- Are changes in the code and intent properly captured in the description of the PR?
- consider appropriateness of tests for implementation details

>   - we want to avoid needing to touch many tests for changes in on internal details

Reviewer 3: - Line 57 invites the question:  should we have a PR creation / review guide for API and another for implementation?

Tests
---------------

- For bug fix PRs: add test which would catch the issue without the fix
- Do newly added tests have the correct granularity?

> - Do tests have a suitable runtime or are unnecessarily large?

Reviewer 3: - Remove words after "runtime"

Reviewer Behavior
-----------------

> - provide timely feedback and respond to changes by the author of the pull request in a reasonable amount of time; it's best to give feedback to pull requests as quickly as possible.

Reviewer 3: - After establishing what PR stands for, use PR throughout (vs. "pull request")

> - only request changes if they are ready to resolve the request upon changes by the author of the pull request; stalling pull requests for requested changes that have been addressed is a problem.

Reviewer 3: - Here, do you mean, "please make sure your PR review comments / change requests are not redundant"

- only review pull requests that have been marked as ready; we have a bunch of pull requests that explore the feasibility of ideas and just need the CI to run. Similarly, pull requests should only be marked as "ready for review" if the author is reasonably happy with the status. If the author mostly seeks feedback on general design and direction, this should be clearly communicated in the pull request description (either "draft" or "ready for review").
- mirror communication with pull request author outside of pull requests (on slack, in person, video calls, etc.) in comments to the pull request.

> - contact authors directly if more clarification is needed.

Reviewer 3: - Are lines 72 and 73 saying the same thing?

> - not be afraid of reviewing pull requests even if they are (slightly) outside their comfort zone.

Reviewer 3: - Maybe also suggest that potential reviewers can partner with an experienced Kokkos team member to perform review to mature reviewing skills?
